### PR TITLE
storage: Ingest snapshot by split

### PR DESCRIPTION
**Pull request by breezewish**
### What problem does this PR solve?

Issue Number: ref #5237

This is one of a series of PRs for https://github.com/pingcap/tiflash/issues/5237, where you can see the overall roadmap.

### What is changed and how it works?

Use LogicalSplitAt + ReplaceData to ingest snapshots when `clear_data_in_range == true`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

**Workload: Ingest 100GB data using Full-Speed, then query a few times**

this PR vs master:

image
image
image
image
image
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Improve ApplySnapshot performance
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/ExternalDTFileInfo.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+struct ExternalDTFileInfo
+{
+    /**
+     * The allocated PageId of the file.
+     */
+    PageId id;
+
+    /**
+     * The handle range of contained data.
```
[the remark this answers is not in the document]
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/ExternalDTFileInfo.h:32]: Actually according to the unit test I see that the handle range would be never nullopt. I'm discovering how to remove this optional wrapper safely. Thanks for pointing it out!

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp [hunk 2]
```diff
@@ -210,6 +140,412 @@ void DeltaMergeStore::ingestFiles(
...
+        else
+        {
+            RUNTIME_CHECK(!external_file.range);
+        }
+
+        files.emplace_back(std::move(file));
+    }
+
+    bool ingest_using_split = false;
+    if (clear_data_in_range && bytes >= 4 * 1024 * 1024)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:494]: maybe reuse `dm_context.delta_small_column_file_bytes` for the threshold is better

**Comment by breezewish** [comment]
Not sure how integration test failed, but looks like the storage meta data is corrupted, which seems to be not related with ingestion:

DB::Exception: Cannot open file /tmp/tiflash/data/db/metadata/db_401/t_398.sql, errno: 2, strerror: No such file or directory:  table name: t_398, table id: 398 

Cannot reproduce the problem locally. Try rerun again.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp [hunk 3]
```diff
@@ -210,6 +140,412 @@ void DeltaMergeStore::ingestFiles(
...
+        /*
+         * The segment and the ingest range is perfectly matched. We can
+         * simply replace all of the data from this segment.
+         *
+         * Example:
+         *    │----------- Segment ----------│
+         *    │-------- Ingest Range --------│
+         */
+        const auto new_segment_or_null = segmentDangerouslyReplaceData(dm_context, segment, file);
+        const bool succeeded = !!new_segment_or_null;
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:391]: Why `!!`?

Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:391]: `new_segment_or_null` is a `std::optional<...>`, so `!!` casts it to boolean unconditionally

Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:391]: maybe use `new_segment_or_null != nullptr`

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp [hunk 4]
```diff
@@ -210,6 +140,412 @@ void DeltaMergeStore::ingestFiles(
...
+         *
+         * We will logical split the segment to form a perfect matching segment:
+         *    │--------------- Segment ------│--------│
+         *    │-------- Ingest Range --------│
+         */
+        const auto [left, right] = segmentSplit(dm_context, segment, SegmentSplitReason::IngestBySplit, ingest_range.end, SegmentSplitMode::Logical);
+        // We don't care whether it is succeeded or not, because the caller always need to retry
+        // from the beginning in this case.
+        UNUSED(left);
+        UNUSED(right);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:409]:
```suggestion
        UNUSED(left, right);
```
`UNUSED` can accept multiple prarms

**Comment by JaySon-Huang** [comment]
We can add some tests that applySnapshot and read/applySnapshot and background segmentMergeDelta run concurrently (maybe in later PR)

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp (hunk 3 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp]: Updated. Thanks for the suggestion!

File: dbms/src/Storages/DeltaMerge/ExternalDTFileInfo.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/ExternalDTFileInfo.h:32]: The optional wrapper has been replaced by some RUNTIME checks.

File: dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp [hunk 5]
```diff
@@ -289,6 +307,42 @@ void SSTFilesToDTFilesOutputStream<ChildStream>::cancel()
...
+    {
+        current_file_range = RowKeyRange( //
+            block_start.toRowKeyValue(),
+            block_end,
+            schema_snap->is_common_handle,
+            schema_snap->rowkey_column_size);
+    }
+    else
+    {
+        RUNTIME_CHECK(compare(block_start, current_file_range->getStart()) > 0);
```
Comment (breezewish) [inline dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:340]: @JaySon-Huang Will we split a key into multiple blocks (due to multiple versions)? If so, then this check may fail I think.

File: dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp [hunk 6]
```diff
@@ -289,6 +307,42 @@ void SSTFilesToDTFilesOutputStream<ChildStream>::cancel()
...
+    const auto rowkey_column = RowKeyColumnContainer(pk_col.column, schema_snap->is_common_handle);
+    auto & current_file_range = ingest_files_range.back();
+
+    auto const block_start = rowkey_column.getRowKeyValue(0);
+    auto const block_end = rowkey_column.getRowKeyValue(pk_col.column->size() - 1) //
+                               .toRowKeyValue()
+                               .toPrefixNext(); // because range is right-open.
+
+    // TODO: what if there are a lof of versions, exceeding a block size? Will our checks fail?
+    RUNTIME_CHECK(compare(block_start, block_end.toRowKeyValueRef()) < 0);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:328]: For rough set filter working correctly, `BoundedSSTFilesToBlockInputStream` ensures that if there are multiple rows with the same rowkey, they must be in the same block.

But we do need to handle that a block that only contains one rowkey. Calling `toPrefixNext` for getting the block_end may be wrong under this situation.

https://github.com/pingcap/tiflash/blob/b232bcd66ccc2684f9b0688daf941832077affd2/dbms/src/Storages/DeltaMerge/SSTFilesToBlockInputStream.h#L115-L117

Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:328]: Interesting, when will we have a block containing only one key? The block size is small, or the versions of key is too much?

Seems that the boundary will be fine, as long as we don't meet one key in multiple blocks.

Consider:

```
[1, 1, 1, 1], [2, ....]
```

The range would be:

```
[1, 2) [2, ...)     (int handle)
[0x1, 0x2) [0x2, ...)  (common handle)
```

Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:328]: When the versions of key is too much(8192 rows per block by default), we will get a block containing only one key.

Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:328]: I get it, `toPrefixNext` solve the problem

File: dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp (hunk 5 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:340]: Never have overlapped ranges, the children `BoundedSSTFilesToBlockInputStreamPtr` ensure that for the correctness of rough set filter in DTFile

**Review by JaySon-Huang — APPROVED** [review]
LGTM

but I think it would be better to let 5879 get merged first. Or it will introduce some conflict for apply snapshot codes when cherry-picking 5879 to previous versions

**Comment by breezewish** [comment]
> LGTM
> 
> but I think it would be better to let 5879 get merged first. Or it will introduce some conflict for apply snapshot codes when cherry-picking 5879 to previous versions

Makes sense. Let's merge #5879 first.

File: dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp (hunk 5 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:340]: Good news. I will add some checks in the next PR along with the test.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp [hunk 7]
```diff
@@ -210,6 +140,401 @@ void DeltaMergeStore::ingestFiles(
...
+     *  │               │-- DMFile --│- DMFile --│---- DMFile ----│- DMFile --│       │
+     *  │- Segment --│-- Seg --│------- Segment -----│- Seg -│------- Segment --------│
+     *
+     * This is what we will ingest:
+     * Iterate 0:
+     * -Inf                                                                         +Inf
+     *  │        │--------------- Ingest Range ---------------│                       │
+     *  │               │-- DMFile --│                                                │
+     *  │            │-- Seg --│------- Segment -----│                                │
+     *                   ↑              ↑ The segment we ingest DMFile into
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:246]:
```suggestion
     *                   ↑                The segment we ingest DMFile into
```

**Review by flowbehappy — APPROVED** [review]
LGTM

**Comment by breezewish** [comment]
@JaySon-Huang @flowbehappy Great thanks for the review! Let me use the latest branch to perform some e2e tests as the final verifications. I will trigger the merge after that.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp (hunk 7 above)
Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp:246]: Oh, may be I should made comments more clear. I would like to illustrate what we will iterate for the outmost loop `for (size_t file_idx = 0; file_idx < files.size(); file_idx++)`.

**Pull request merged**